modules/calcite/src/test/java/org/apache/ignite/internal/processors/query/calcite/integration/TableDdlIntegrationTest.java

```diff
@@ -995,6 +1016,51 @@ public void testPrimaryKeyInlineSize() {
             "MY_TABLE", 5 + 3);
     }
 
+    /** */
+    @Test
+    public void testNonPersistentRejoinsWithDynamicTablesOverPredefinedCaches() throws Exception {
+        stopAllGrids();
+
+        try {
+            persistence = false;
+
+            CacheConfiguration<?, ?> cacheCfg = new CacheConfiguration<>("TEST_CACHE")
+                .setBackups(1)
+                .setCacheMode(CacheMode.PARTITIONED)
+                .setAtomicityMode(CacheAtomicityMode.TRANSACTIONAL)
+                .setWriteSynchronizationMode(CacheWriteSynchronizationMode.PRIMARY_SYNC);
+
+            cacheConfigurations = new CacheConfiguration<?, ?>[] {cacheCfg};
+
+            client = startGrids(3);
+
+            sql("CREATE TABLE IF NOT EXISTS TEST_TBL(ID INTEGER PRIMARY KEY, VAL VARCHAR) WITH \"CACHE_NAME=TEST_CACHE\"");
+
+            assertEquals(0, sql("SELECT * FROM TEST_TBL").size());
+
+            int testGrid = 2;
+
+            stopGrid(testGrid);
+            startGrid(testGrid);
+
+            awaitPartitionMapExchange();
+
+            for (int i = 0; i < 100; ++i)
+                assertEquals(1, sql("INSERT INTO TEST_TBL VALUES(" + (i + 1) + ", '" + (i + 1000) + "')").size());
+
+            assertEquals(100, grid(testGrid).cache("TEST_CACHE").size());
+
+            assertEquals(100, sql("SELECT * FROM TEST_TBL").size());
+        }
+        finally {
+            persistence = true;
+
+            afterTestsStopped();
+
+            beforeTestsStarted();
```

Comment:
Looks weird, let's introduce new test class for this test
Also:
- the same problem can be reproduced on H2 engine.
- the same problem exists for persistent (if we clean persistent dir for testGrid after stopping)